### ISSUE TITLE
Fix scripted debian repository signing

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#2721]
## Background

Our [infrastructure](https://github.com/QubesOS/qubes-infrastructure/) allow to migrate packages from testing to stable repository using signed comment in [updates-status repository](https://github.com/QubesOS/updates-status/issues). This works well for Fedora repositories, where only packages are signed, but repository metadata not - so migrating a package doesn't need access to a signing key.
For Debian repositories, it's another story - only repository metadata is signed, so both uploading new package and migrating package require access to a signing key.

Current configuration (linked above) give access to a signing key only after uploading build log. The goal of such configuration is to have some trace what was signed, and even if a build environment got compromised, have (public) evidence that it happened.
This goal is achieved only partially:
 - there is no way that key-holding-VM know that what it gets for signing really corresponds to the log just uploaded; this problem is hard on multiple levels, and in case of Debian it's even harder, as the thing being signed (Release file) is only indirectly connected to a package (contains a hash or a file (Packages) containing a hash of actual package...).
 - uploading a log give access to a signing key for 5 minutes, without any limit on usage count

But this is still better than giving access to a signing key, without enforcing any trace.

## Problem

When just moving packages between repositories, nothing is built. In fact, output from `make update-repo-current` is also submitted as a "build log", but when it needs an access to a signing key, the log is still not finished. This means that moving Debian package from testing to stable by a comment in updates-status repository does not work without a manual intervention.

**Comment by DemiMarie** [comment]
This can be fixed by splitting the output of `make update-repo-current` into two logs:

- Log 1 consists of everything *before* the file needs to be signed.
- Log 2 consists of everything *after* the signing step.

**Comment by DemiMarie** [comment]
By running each build in a DispVM with no network or qrexec access, we can ensure that the only data that the build could have used is the data sent to it in the build call.

**Comment by marmarek** [comment]
> By running each build in a DispVM with no network or qrexec access, we can ensure that the only data that the build could have used is the data sent to it in the build call.

Yes, that is also on the roadmap. But for a slightly different reason - to isolate components builds from each other (be sure that none of one components sources could compromise other components sources). Basically to allow assigning maintainers to separate components, and not give them in practice control over everything. The prime blocker for this is Makefile.builder, which is basically executed outside of any sandbox. Some details at https://github.com/QubesOS/qubes-builder/pull/123#issuecomment-697010703

As for the network and qrexec access - this is already in place - both mock and pbuilder use namespaces to isolate the build environment, especially a network namespace without any connection to the outside. This is only as strong as namespaces (still a common kernel), but gives assurance about using only the right build inputs.

**Comment by marmarek** [comment]
> This can be fixed by splitting the output of `make update-repo-current` into two logs:
> 
>     * Log 1 consists of everything _before_ the file needs to be signed.
> 
>     * Log 2 consists of everything _after_ the signing step.

Yes, this sounds like a good approach. And the Log 1 should include information about what to be signed, preferably a diff of changes.

At some point we could strengthen signature key control by requiring the log to explicitly include hashes that are going to be signed (not just "anything that is is requested in the next few minutes"). But that is a separate task, not required right now.

**Comment by DemiMarie** [comment]
This should be P: blocker, not P: minor.  @andrewdavidwong can you make me part of the @QubesOS organization?

**Comment by andrewdavidwong** [comment]
> This should be P: blocker, not P: minor, as it blocks #6177.

Updated, thanks!

> @andrewdavidwong can you make me part of the @QubesOS organization?

Looks like someone already did this. Welcome! :slightly_smiling_face:
